text/0000-node-api-borrow.md

```diff
@@ -0,0 +1,459 @@
+}
+```
+
+The `ResultExt` trait is identical to `JsResultExt` except that it does not require the `Ok` branch to implement `Value`. We will most likely want to deprecate `JsResultExt` in the future since this trait is strictly more powerful.
+
+### `Lock` extension for `Borrow`
+
+While the `Borrow` trait provides the required functionality, it requires that the trait be in scope. In order to make this more ergonomic for users, the dynamically checked methods are mirrored on `Lock`. These methods do *not* require the trait in scope.
+
+```rust
+impl<'a: 'cx, 'cx, C> Lock<'cx, C>
+    where
+        C: Context<'a>,
+{
+    /// Constructs a new [`Lock`] and locks the VM. See also [`Context::lock`].
+    pub fn new(cx: &'cx mut C) -> Lock<'cx, C> {
+        Lock {
+            cx,
+            ledger: Default::default(),
+        }
+    }
+
+    /// Dynamically checked immutable borrow.
+    ///
+    /// See [`Borrow::try_borrow`].
+    pub fn try_borrow<T>(&self, buf: &T) -> Result<Ref<T::Item>, BorrowError>
+        where
+            T: Borrow,
+    {
+        buf.try_borrow(self)
+    }
+
+    /// Dynamically checked mutable borrow.
+    ///
+    /// See [`Borrow::try_borrow_mut`].
+    pub fn try_borrow_mut<T>(&self, buf: &mut T) -> Result<RefMut<T::Item>, BorrowError>
+        where
+            T: Borrow,
+    {
+        buf.try_borrow_mut(self)
+    }
+}
+```
+
+#### Typed Arrays
+
+`TypedArray` was introduced in ECMAScript 2015 and defines a [typed _view_](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray) into an `ArrayBuffer`.
+
+While `ArrayBuffer` is always a chunk of `u8`, views may have different numeric parameters (`Int8Array`, `Uint8Array`, `Uint8ClampedArray`, `Int16Array`, `Uint16Array`, `Int32Array`, `Uint32Array`, `Float32Array`, `Float64Array`, `BigInt64Array`, `BigUint64Array`)
+
+The specific type of typed array is indicated by a type parameter on `JsTypeArray`.
+
+```rust
+pub struct JsTypedArray<T> {
+    local: raw::Local,
+    _type: PhantomData<T>,
+}
+
+impl JsTypedArray<i8> {}
+impl JsTypedArray<u8> {}
+impl JsTypedArray<i16> {}
+impl JsTypedArray<u16> {}
+impl JsTypedArray<i32> {}
+impl JsTypedArray<u32> {}
+impl JsTypedArray<f32> {}
+impl JsTypedArray<f64> {}
+impl JsTypedArray<i64> {}
+impl JsTypedArray<u64> {}
+```
+
+Both `Uint8Array` and `Uint8ClampedArray` are represented with a type parameter of `u8`. "Clamped" is not represented in this design because it could not be ergonomically enforced.
+
+##### Borrow
+
+It is possible to implement `Borrow` generically for all` JsTypedArray`.
+
+```rust
+impl<T: Copy> Borrow for JsTypedArray<T> {
+    type Item = T;
+
+    /* ... */
+}
+```
+
+However, since `JsTypedArray` may overlap with each other, it is important that `Lock` properly track the _region_ of memory and not only the starting pointer. `Lock` will maintain a set of pointer `Range` to track which buffers are currently borrowed.
+
+### Module Layout
```

Comment:
We had a great conversation today about the naming and layout. Here were the main design constraints we identified:

- Commonly used JS types should be in `neon::types` so there's one main place to look in the docs to understand how to use standard JS types.
- Rarely-used datatypes that are specific to typed arrays should not pollute `neon::types` so that the experience of reading the docs for standard JS types isn't overwhelming and full of rarely-needed domain-specific details.
- Anything that users are likely to need to bring into scope for common use cases should not have a name that conflicts with a common Rust standard library.
- Concepts analogous to the Rust standard library should nevertheless use naming conventions that are suggestive of those analogies, to help learnability.
- Neon modules should have names that match normal intuitions as users are likely to think about them, not abstract or theoretical concepts.
- Types that are part of generic Neon functionality (e.g. contexts) should not have functionality specific to typed arrays.
- "Punning" inherent methods after trait methods is useful for making common, idiomatic patterns more ergonomic; the extra boilerplate in the implementation is worth it for the win to user experience.
- Abstractions relating to typed arrays should not have overly generic names.
- The commonly-used name for the typed arrays space is "typed arrays" :)
- The common intuition for the kind of data structure in this space is the "buffer" (`Buffer`, `ArrayBuffer`, `SharedArrayBuffer`)
- Module names that are a noun should be singular since this seems to be idiomatic in the Rust ecosystem (`types` was a special case since `type` is a reserved word)

With that, the ideas we came up with for the naming and layout were:

- Move `Lock` into `neon::context` since it's a generic concept relating to the state of the VM
  - Remove the inherent methods relating to typed arrays, especially since we want to encourage `buffer.borrow()` instead of `lock.borrow()` as the idiomatic pattern.
- Rename `Borrow` to `TypedArray` and `BorrowError` to `TypedArrayError`
- Add inherent methods to the typed array types so users don't need to import `TypedArray` unless they're writing abstractions that need a generic trait
- Keep `JsArrayBuffer`, `JsBuffer`, and `JsTypedArray` in `neon::types`
- Keep `Ref` and `RefMut` named as is and in `neon::types::binary`
- Rename `neon::types::binary` to `neon::types::buffer`

Separately, we also discussed the idea of moving all error types into `neon::result` and renaming that module to `neon::error`, but that's out of scope for this RFC. Since it's a backwards-incompatible change, we should tackle that soon and avoid releasing this (also backwards-incompatible) RFC until we release that one as well, to avoid the number of incompatible upgrades users need to go through. But still it's good to land the PRs individually so we'll separate that into a followup.

Reply:
> Add inherent methods to the typed array types so users don't need to import `TypedArray` unless they're writing abstractions that need a generic trait

I don't feel as confident about this one. It felt like a good win when the name conflicted with `std::borrow::Borrow`, but renaming it to `TypedArray` resolves that issue. It doesn't feel super rusty to duplicate these.

We could add them later with a semver compatible change. What do you think about deferring it for now?

Reply:
The other change I think is to leave `BorrowError` named that instead of `TypedArrayError`. It's unclear what that error is and the error type is scoped to the `buffer` module.